### ISSUE TITLE
Cleanup of reserved tokens (releaseToken).

### ISSUE THREAD

**Issue by boaks** [eclipse/californium#232]
BaseMatcher:

>        public void cancelObserve(final byte[] token) {
> 		// we do not know the destination endpoint the requests have been sent
> 		// to therefore we need to find them by token only
> 		for (Exchange exchange : exchangeStore.findByToken(token)) {
> 			exchange.getRequest().cancel();
> 			KeyToken idByToken = KeyToken.fromOutboundMessage(exchange.getCurrentRequest());
> 			exchangeStore.releaseToken(idByToken);
> 		}
> 		observationStore.remove(token);
> 	}

Though Exchanges for observe are not longer "longterm" stored in the exchangeStore, this would not find the Exchange. Using observationStore for lookup the Exchange to create the KeyToken, will depend on the remove order. 

So I would go for using the Exchange itself.
Any comments?

**Comment by sophokles73** [comment]
I am sorry, but it is unclear to me what you want to do/change. I do understand that it is hard to get a grip on the `Exchange` instance that originally created the observation, right? But I do not understand what you are proposing to do here?

**Comment by boaks** [comment]
I want to change the signature of cancel. e.g.

> public void cancelObserve(Exchange exchange)

or something similar. But I'm currently not working on it.

**Comment by sbernard31** [comment]
+1 with @sophokles73, Why should we do that ?

Do you want to change the signature of CoapEndpoint.cancelObservation() too ?

**Comment by sbernard31** [comment]
The idea of searching in exchangeStore is only if there is a notification using block2 in progress.

**Comment by boaks** [comment]
As far as I understood the design, a observe exchange is only stored in the exchangeStore using the KeyToken for the initial request. But the token is maintained in the TokenProvider even after the Exchange is completed. So, without a Exchange stored by KeyToken, how should the token be released in the TokenProvider?

But in general I think, we should add more cyclic logging information to see, if there are real leaks in the java heaps or just in the understanding of implementation :-).

**Comment by sbernard31** [comment]
I think you're right there is an issue with this part of the code.
We should always call `releaseToken()`.
I look at this history and originally (ecb36f7ea4059897c85983efb1709a7750e01d55) the code looks like that :
```java
@Override
public void cancelObserve(final byte[] token) {
	// search for pending blockwise exchange for this observe request
	for (Entry<KeyToken, Exchange> key : exchangesByToken.entrySet()) {
		Request cachedRequest = key.getValue().getRequest();
		if (cachedRequest != null && Arrays.equals(token, cachedRequest.getToken())) {
			cachedRequest.cancel();
		}
	}
	observationStore.remove(token);
	tokenProvider.releaseToken(token);
}	
```
I don't know exactly why this was changed.

I agree about leaks and logging. Maybe we could also add some [memory leaks tests](https://github.com/eclipse/californium/blob/2.0.x/californium-core/src/test/java/org/eclipse/californium/core/test/MemoryLeakingHashMapTest.java))

**Comment by boaks** [comment]
I guess, the issue for changing that was #173

**Comment by sophokles73** [comment]
The token is released by the `exchangeStore.releaseToken(idByToken)` call.
I do not think that there is an issue here. Is there a test case available illustrating the (assumed) problem?

**Comment by boaks** [comment]
I would say, the would be released, if `exchangeStore.releaseToken(idByToken)` would be called.
But, as mentioned in #226, I can't see, that it is called.

Yes, test case would be perfect. But though some changes for test are already pending, it will take some time.

**Comment by sophokles73** [comment]
Ok, I'll try to sum up what I think to have understood your point to be:

When client code uses a token to cancel an observation using `Endpoint.cancelObservation()`, we will not be able to find the corresponding `Exchange` in the `MessageExchangeStore` because the original exchange for establishing the observe relation will most likely have completed already and thus been removed from the exchange store. You therefore propose to try to get hold of the original `Exchange` by looking it up in the `ObservationStore`, because this is where we keep track of the original request that had established the observation, right?
However, IMHO this will also not be sufficient because we have scoped the tokens to endpoint addresses with the introduction of the `MessageExchangeStore` and we could therefore end up with multiple Observations for different peer addresses being tracked in the `ObservationStore` using the same token. So I guess we will need to add the peer address as a parameter to `Endpoint.cancelObservation()`, don't you think?

**Comment by boaks** [comment]
That's what I ment with in my second comment.

> I want to change the signature of cancel. e.g.

I didn't saw a easy solution, therefore I opened an issue as reminder to work later on it.

**Comment by sbernard31** [comment]
>So I guess we will need to add the peer address as a parameter to Endpoint.cancelObservation(), don't you think?

Using peer address is not enough because, Observation is longterm relation, so we need a longterm identifier for the peer. You know the problem with NAT.
You know [my position](https://mailarchive.ietf.org/arch/search/?email_list=core&gbt=1&index=0vwuEnq6IRHJc70l3BTggX6p9vs) on this. I think we should use the DTLS identity or the session if there is no authentication.(see #173)

**Comment by sophokles73** [comment]
Oh man, it feels like going around in circles ...
So, it looks like we first need to change the way we identify endpoints within Scandium/Californium, right?
Once we have that in place everything else will follow naturally ;-)

**Comment by boaks** [comment]
> Oh man, it feels like going around in circles ...

:-)

My impression was also, that it's not too easy.

**Comment by sbernard31** [comment]
>So, it looks like we first need to change the way we identify endpoints within Scandium/Californium, right?

It sounds a good plan.

**Comment by boaks** [comment]
Closed with PR #521
